### PR TITLE
RS/YJ/Rule 23-16

### PR DESCRIPTION
**Pull request by yunjoonjung-PNNL**
This PR includes Rule23-16 development. Thanks!

### CODE REVIEW

File: rct229/rules/section23/section23rule16.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+        hot_water_loop_type_dict = {}
+        for preheat_system in find_all("$..preheat_system", rmi_b):
+            hot_water_loop = preheat_system["hot_water_loop"]
+            hot_water_loop_type_dict[hot_water_loop] = find_exactly_one_fluid_loop(
+                rmi_b, hot_water_loop
+            )["type"]
+
+        baseline_system_types_dict = get_baseline_system_types(rmi_b)
+        applicable_hvac_sys_ids = [
```
Comment (weilixu) [inline rct229/rules/section23/section23rule16.py:95]: Same comment as 23-2 here

File: rct229/rules/section23/section23rule16.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+APPLICABLE_SYS_TYPES = [
+    HVAC_SYS.SYS_5,
+    HVAC_SYS.SYS_6,
+    HVAC_SYS.SYS_7,
+    HVAC_SYS.SYS_8,
+]
+
+HEATING_SYSTEM = schema_enums["HeatingSystemOptions"]
+Fluid_Loop = schema_enums["FluidLoopOptions"]
```
Comment (weilixu) [inline rct229/rules/section23/section23rule16.py:27]: Suggest to change this to `FLUID_LOOP`

**Review by weilixu — APPROVED** [review]
Approved with some stylish comments

File: rct229/rules/section23/section23rule16.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    def create_data(self, context, data):
+        rmi_b = context.baseline
+
+        hvac_max_zone_setpoint_dict = {}
+        for zone in find_all("$..zones[*]", rmi_b):
+            zone_design_heating_setpoint = getattr_(
+                zone, "dsgn_heating_setpoint", "design_thermostat_heating_setpoint"
+            )
+            for terminal in find_all("$..terminals[*]", zone):
+                terminal_system = getattr_(
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:69]: Fixed.

File: rct229/rules/section23/section23rule16.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            ]
+        )
+
+    def create_data(self, context, data):
+        rmi_b = context.baseline
+
+        hvac_max_zone_setpoint_dict = {}
+        for zone in find_all("$..zones[*]", rmi_b):
+            zone_design_heating_setpoint = getattr_(
+                zone, "dsgn_heating_setpoint", "design_thermostat_heating_setpoint"
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:66]: Fixed.

File: rct229/rules/section23/section23rule16.py (hunk 1 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:95]: Fixed.

File: rct229/rules/section23/section23rule16.py (hunk 2 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:27]: Fixed.

File: rct229/rules/section23/section23rule16.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                        hvac_max_zone_setpoint_dict[terminal_system]
+                        < zone_design_heating_setpoint
+                    ):
+                        hvac_max_zone_setpoint_dict[
+                            terminal_system
+                        ] = zone_design_heating_setpoint
+                else:
+                    hvac_max_zone_setpoint_dict[
+                        terminal_system
+                    ] = zone_design_heating_setpoint
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:85]: This is brilliant. Thank you Weili!

File: rct229/rules/section23/section23rule16.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        ]
+
+        return {
+            "hvac_max_zone_setpoint_dict": hvac_max_zone_setpoint_dict,
+            "hot_water_loop_type_dict": hot_water_loop_type_dict,
+            "applicable_hvac_sys_ids": applicable_hvac_sys_ids,
+        }
+
+    def list_filter(self, context_item, data):
+        applicable_hvac_sys_ids = data["applicable_hvac_sys_ids"]
```
Comment (charliepnnl) [inline rct229/rules/section23/section23rule16.py:111]: I prefer creating a variable for the context_item as the first line of code inside a list_filter:
```
hvac_sys_b = context_item.baseline
```

**Review by charliepnnl — APPROVED** [review]
Approved with some comments that should be addressed.
Also, the RDS is confusing at best and could use some restructuring.

File: rct229/rules/section23/section23rule16.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+APPLICABLE_SYS_TYPES = [
+    HVAC_SYS.SYS_5,
+    HVAC_SYS.SYS_6,
+    HVAC_SYS.SYS_7,
+    HVAC_SYS.SYS_8,
+]
+
+HEATING_SYSTEM = schema_enums["HeatingSystemOptions"]
+Fluid_Loop = schema_enums["FluidLoopOptions"]
+TENP_20_degF = 20.0 * ureg("delta_degF")
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:28]: Great point! Fixed.

File: rct229/rules/section23/section23rule16.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py]: Thanks for letting me know about this! Fixed.

File: rct229/rules/section23/section23rule16.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+        hvac_max_zone_setpoint_dict = {}
+        for zone in find_all("$..zones[*]", rmi_b):
+            zone_design_heating_setpoint = getattr_(
+                zone, "dsgn_heating_setpoint", "design_thermostat_heating_setpoint"
+            )
+            for terminal in find_all("$..terminals[*]", zone):
+                terminal_system = getattr_(
+                    terminal,
+                    "system",
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:72]: Fixed.

File: rct229/rules/section23/__init__.py [hunk 9]
```diff
@@ -7,7 +7,7 @@
     # "section23rule6",
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/__init__.py:7]: Great point. This should be uncommented.

File: rct229/utils/pint_utils.py [hunk 10]
```diff
@@ -65,6 +65,8 @@ class ZERO:
     UA = U_FACTOR * AREA
     FLOW = VOLUME / ureg("minute")
 
+    TEMPERATURE = 0 * ureg("degC")
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/utils/pint_utils.py:68]: Great point! Thanks for pointing this out. Fixed.

File: rct229/rules/section23/section23rule16.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+                        hvac_max_zone_setpoint_dict.get(hvac_id, ZERO.TEMPERATURE),
+                    )
+                    for hvac_id in find_all(
+                        "$..terminals[*].served_by_heating_ventilating_air_conditioning_system",
+                        zone,
+                    )
+                }
+
+        # find preheat_system's hot water loop type
+        hot_water_loop_type_dict = {}
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:87]: Fixed.

File: rct229/rules/section23/section23rule16.py (hunk 6 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section23/section23rule16.py:111]: Fixed.

**Comment by yunjoonjung-PNNL** [comment]
Note)
I'll merge this PR first and revisit all the section 23 RDS to see whether there is any confusing part. If so, I'll create a separate PR for updating the RDS.

**Pull request merged**